{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "df63f901",
   "metadata": {},
   "source": [
    "<H1>SCN2A: Generate Transcript HGVS </H1>\n",
    "<p>In this notebook, we generate nucleotide HGVS expressions for the variants listed in \n",
    "  <a href=\"https://pubmed.ncbi.nlm.nih.gov/33731876/\" target=\"__blank\">Crawford et al, Computational analysis of 10,860 phenotypic annotations in individuals with SCN2A-related disorders</a>. In the Supplementary Table of this publication, variants were given in Protein notation. However, for our software, we require transcript level HGVS notations. We used the <a href=\"http://useast.ensembl.org/Homo_sapiens/Tools/VR\" target=\"__blank\">Variant Recoder</a> tool of Ensembl to extract the corresponding data and here create a mapping file that will be used in the other SCN2A notebook to generate phenopackets.</p>\n",
    "  <p>Note that <a href=\"https://useast.ensembl.org/Homo_sapiens/Transcript/Summary?db=core;g=ENSG00000136531;r=2:165194993-165392310;t=ENST00000375437\" target=\"__blank\">ENST00000375437.7</a> is the MANE select transcript that we use here. It corresponds to <a href=\"https://www.ncbi.nlm.nih.gov/nuccore/NM_001040142.2\" target=\"__blank\">NM_001040142.2</a>.</p>\n",
    " <p>The purpose of this notebook is to prepare HGVS-compliant variant strings that we will use in the Crawford_SCN2A notebook to generate phenopackets.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9e5d76dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8f701745",
   "metadata": {},
   "outputs": [],
   "source": [
    "vr_df = pd.read_csv(\"input/vr_output\", comment='#', delimiter='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f43cfa5f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uploaded_variant</th>\n",
       "      <th>Allele</th>\n",
       "      <th>HGVSg</th>\n",
       "      <th>HGVSc</th>\n",
       "      <th>HGVSp</th>\n",
       "      <th>VARID</th>\n",
       "      <th>Variant_synonyms</th>\n",
       "      <th>MANE_Select</th>\n",
       "      <th>VCF</th>\n",
       "      <th>HGVSg.1</th>\n",
       "      <th>HGVSc.1</th>\n",
       "      <th>HGVSp.1</th>\n",
       "      <th>VARID.1</th>\n",
       "      <th>MANE_Select.1</th>\n",
       "      <th>VCF.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSP00000364586:p.V423L</td>\n",
       "      <td>T</td>\n",
       "      <td>NC_000002.12:g.165313992G&gt;T</td>\n",
       "      <td>ENST00000283256.10:c.1267G&gt;T, ENST00000375437....</td>\n",
       "      <td>ENSP00000283256.6:p.Val423Leu, ENSP00000364586...</td>\n",
       "      <td>CM176086</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_000002.12:g.165313992G&gt;T;ENST00000375437.7:...</td>\n",
       "      <td>2-165313992-G-T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSP00000364586:p.L501*</td>\n",
       "      <td>TA</td>\n",
       "      <td>NC_000002.12:g.165315588_165315589delinsTA</td>\n",
       "      <td>ENST00000283256.10:c.1501_1502delinsTA, ENST00...</td>\n",
       "      <td>ENSP00000283256.6:p.Leu501Ter, ENSP00000364586...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_000002.12:g.165315588_165315589delinsTA;ENS...</td>\n",
       "      <td>2-165315588-CT-TA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSP00000364586:p.R583*</td>\n",
       "      <td>T</td>\n",
       "      <td>NC_000002.12:g.165323231C&gt;T</td>\n",
       "      <td>ENST00000283256.10:c.1747C&gt;T, ENST00000375437....</td>\n",
       "      <td>ENSP00000283256.6:p.Arg583Ter, ENSP00000364586...</td>\n",
       "      <td>rs1553571901, CM154971, COSV51841058</td>\n",
       "      <td>dbSNP HGVS: NP_001035232.1:p.Arg583Ter,NM_0010...</td>\n",
       "      <td>NC_000002.12:g.165323231C&gt;T;ENST00000375437.7:...</td>\n",
       "      <td>2-165323231-C-T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSP00000364586:p.D609*</td>\n",
       "      <td>TAA</td>\n",
       "      <td>NC_000002.12:g.165323309_165323311delinsTAA</td>\n",
       "      <td>ENST00000283256.10:c.1825_1827delinsTAA, ENST0...</td>\n",
       "      <td>ENSP00000283256.6:p.Asp609Ter, ENSP00000364586...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_000002.12:g.165323309_165323311delinsTAA;EN...</td>\n",
       "      <td>2-165323309-GAC-TAA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSP00000364586:p.D609*</td>\n",
       "      <td>TAG</td>\n",
       "      <td>NC_000002.12:g.165323309_165323311delinsTAG</td>\n",
       "      <td>ENST00000283256.10:c.1825_1827delinsTAG, ENST0...</td>\n",
       "      <td>ENSP00000283256.6:p.Asp609Ter, ENSP00000364586...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_000002.12:g.165323309_165323311delinsTAG;EN...</td>\n",
       "      <td>2-165323309-GAC-TAG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Uploaded_variant Allele  \\\n",
       "0  ENSP00000364586:p.V423L      T   \n",
       "1  ENSP00000364586:p.L501*     TA   \n",
       "2  ENSP00000364586:p.R583*      T   \n",
       "3  ENSP00000364586:p.D609*    TAA   \n",
       "4  ENSP00000364586:p.D609*    TAG   \n",
       "\n",
       "                                         HGVSg  \\\n",
       "0                  NC_000002.12:g.165313992G>T   \n",
       "1   NC_000002.12:g.165315588_165315589delinsTA   \n",
       "2                  NC_000002.12:g.165323231C>T   \n",
       "3  NC_000002.12:g.165323309_165323311delinsTAA   \n",
       "4  NC_000002.12:g.165323309_165323311delinsTAG   \n",
       "\n",
       "                                               HGVSc  \\\n",
       "0  ENST00000283256.10:c.1267G>T, ENST00000375437....   \n",
       "1  ENST00000283256.10:c.1501_1502delinsTA, ENST00...   \n",
       "2  ENST00000283256.10:c.1747C>T, ENST00000375437....   \n",
       "3  ENST00000283256.10:c.1825_1827delinsTAA, ENST0...   \n",
       "4  ENST00000283256.10:c.1825_1827delinsTAG, ENST0...   \n",
       "\n",
       "                                               HGVSp  \\\n",
       "0  ENSP00000283256.6:p.Val423Leu, ENSP00000364586...   \n",
       "1  ENSP00000283256.6:p.Leu501Ter, ENSP00000364586...   \n",
       "2  ENSP00000283256.6:p.Arg583Ter, ENSP00000364586...   \n",
       "3  ENSP00000283256.6:p.Asp609Ter, ENSP00000364586...   \n",
       "4  ENSP00000283256.6:p.Asp609Ter, ENSP00000364586...   \n",
       "\n",
       "                                  VARID  \\\n",
       "0                              CM176086   \n",
       "1                                     -   \n",
       "2  rs1553571901, CM154971, COSV51841058   \n",
       "3                                     -   \n",
       "4                                     -   \n",
       "\n",
       "                                    Variant_synonyms  \\\n",
       "0                                                  -   \n",
       "1                                                  -   \n",
       "2  dbSNP HGVS: NP_001035232.1:p.Arg583Ter,NM_0010...   \n",
       "3                                                  -   \n",
       "4                                                  -   \n",
       "\n",
       "                                         MANE_Select                  VCF  \\\n",
       "0  NC_000002.12:g.165313992G>T;ENST00000375437.7:...      2-165313992-G-T   \n",
       "1  NC_000002.12:g.165315588_165315589delinsTA;ENS...    2-165315588-CT-TA   \n",
       "2  NC_000002.12:g.165323231C>T;ENST00000375437.7:...      2-165323231-C-T   \n",
       "3  NC_000002.12:g.165323309_165323311delinsTAA;EN...  2-165323309-GAC-TAA   \n",
       "4  NC_000002.12:g.165323309_165323311delinsTAG;EN...  2-165323309-GAC-TAG   \n",
       "\n",
       "   HGVSg.1  HGVSc.1  HGVSp.1  VARID.1  MANE_Select.1  VCF.1  \n",
       "0      NaN      NaN      NaN      NaN            NaN    NaN  \n",
       "1      NaN      NaN      NaN      NaN            NaN    NaN  \n",
       "2      NaN      NaN      NaN      NaN            NaN    NaN  \n",
       "3      NaN      NaN      NaN      NaN            NaN    NaN  \n",
       "4      NaN      NaN      NaN      NaN            NaN    NaN  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vr_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d159566b",
   "metadata": {},
   "source": [
    "<h3>Variant Record Mappings</h3>\n",
    "<p>The following code creates a mapping such as D195G: NM_001040142.2:c.584A>G</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f87b4f5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_to_NM_MANE_d = defaultdict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f56d5309",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We extracted 243 mappings\n"
     ]
    }
   ],
   "source": [
    "for _, row in vr_df.iterrows():\n",
    "    # Remove the prefix, i.e., ENSP00000364586:p.\n",
    "    # this will regenerate the String used in the input file\n",
    "    uploaded_var = row['Uploaded_variant'][18:]  \n",
    "    mane_fields = row['MANE_Select'].split(\";\")\n",
    "    nm_mane_filter_obj = filter(lambda t: t.startswith(\"NM_001040142\"), mane_fields)\n",
    "    nm_mane = list(nm_mane_filter_obj)[0]\n",
    "    nm_mane = nm_mane.replace(\"NM_001040142.2:\",\"\")\n",
    "    if nm_mane is None:\n",
    "        raise ValueError(f\"Could not find NM_ for {mane_fields}\")\n",
    "    protein_to_NM_MANE_d[uploaded_var] = nm_mane\n",
    "print(f\"We extracted {len(protein_to_NM_MANE_d)} mappings\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f01328b",
   "metadata": {},
   "source": [
    "<h3>Input the Supplemental file</h3>\n",
    "<p>Only some of the input variants are missense. We need to manually map the remaining variants.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ea0b1a6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"input/41436_2021_1120_MOESM2_ESM.csv\", delimiter='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d26f812f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>famID</th>\n",
       "      <th>broad_phx</th>\n",
       "      <th>variant</th>\n",
       "      <th>variant_type_1</th>\n",
       "      <th>variant_type_2</th>\n",
       "      <th>location</th>\n",
       "      <th>domain</th>\n",
       "      <th>segment</th>\n",
       "      <th>pmid/pmcid</th>\n",
       "      <th>HPO</th>\n",
       "      <th>hpo_term</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>fam1</td>\n",
       "      <td>benign</td>\n",
       "      <td>A202V</td>\n",
       "      <td>missense</td>\n",
       "      <td>missense</td>\n",
       "      <td>Helical Repeat I</td>\n",
       "      <td>DI</td>\n",
       "      <td>S3</td>\n",
       "      <td>28379373</td>\n",
       "      <td>HP:0007359</td>\n",
       "      <td>Focal-onset seizure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>fam1</td>\n",
       "      <td>benign</td>\n",
       "      <td>A202V</td>\n",
       "      <td>missense</td>\n",
       "      <td>missense</td>\n",
       "      <td>Helical Repeat I</td>\n",
       "      <td>DI</td>\n",
       "      <td>S3</td>\n",
       "      <td>28379373</td>\n",
       "      <td>HP:0002069</td>\n",
       "      <td>Generalized tonic-clonic seizures</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>fam1</td>\n",
       "      <td>benign</td>\n",
       "      <td>A202V</td>\n",
       "      <td>missense</td>\n",
       "      <td>missense</td>\n",
       "      <td>Helical Repeat I</td>\n",
       "      <td>DI</td>\n",
       "      <td>S3</td>\n",
       "      <td>28379373</td>\n",
       "      <td>NP:0003808</td>\n",
       "      <td>No Abnormal muscle tone</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>fam1</td>\n",
       "      <td>benign</td>\n",
       "      <td>A202V</td>\n",
       "      <td>missense</td>\n",
       "      <td>missense</td>\n",
       "      <td>Helical Repeat I</td>\n",
       "      <td>DI</td>\n",
       "      <td>S3</td>\n",
       "      <td>28379373</td>\n",
       "      <td>NP:0012443</td>\n",
       "      <td>No Abnormality of brain morphology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>fam1</td>\n",
       "      <td>benign</td>\n",
       "      <td>A202V</td>\n",
       "      <td>missense</td>\n",
       "      <td>missense</td>\n",
       "      <td>Helical Repeat I</td>\n",
       "      <td>DI</td>\n",
       "      <td>S3</td>\n",
       "      <td>28379373</td>\n",
       "      <td>NP:0000717</td>\n",
       "      <td>No Autism</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  famID broad_phx variant variant_type_1 variant_type_2          location  \\\n",
       "0  fam1    benign   A202V       missense       missense  Helical Repeat I   \n",
       "1  fam1    benign   A202V       missense       missense  Helical Repeat I   \n",
       "2  fam1    benign   A202V       missense       missense  Helical Repeat I   \n",
       "3  fam1    benign   A202V       missense       missense  Helical Repeat I   \n",
       "4  fam1    benign   A202V       missense       missense  Helical Repeat I   \n",
       "\n",
       "  domain segment pmid/pmcid         HPO                            hpo_term  \n",
       "0     DI      S3   28379373  HP:0007359                 Focal-onset seizure  \n",
       "1     DI      S3   28379373  HP:0002069   Generalized tonic-clonic seizures  \n",
       "2     DI      S3   28379373  NP:0003808             No Abnormal muscle tone  \n",
       "3     DI      S3   28379373  NP:0012443  No Abnormality of brain morphology  \n",
       "4     DI      S3   28379373  NP:0000717                           No Autism  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "19c78b83",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We extracted 414 families\n"
     ]
    }
   ],
   "source": [
    "fam_to_var_d = defaultdict()\n",
    "for _, row in df.iterrows():\n",
    "    famID = row[\"famID\"]\n",
    "    variant = row[\"variant\"]\n",
    "    fam_to_var_d[famID] = variant\n",
    "print(f\"We extracted {len(fam_to_var_d)} families\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7b871f4",
   "metadata": {},
   "source": [
    "<h2>Extracting variants that could not be mapped by Variant Report</h2>\n",
    "<p>The variants in <tt>manual_lookups</tt> were looked up manually in the original publications.</p>\n",
    "<p>The variants in the <tt>malformed</tt> were discarded since the nucleotide/amino acid changes because\n",
    "information about the nuceltodie changes could not be found or because the variants\n",
    "do not match the prefered transcript and\n",
    "no transcript was provided in the original publication.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3260714d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# These are based on NM_001040142.2\n",
    "manual_lookups = {\n",
    "\n",
    "   \"A202V\": [\"c.605C>T\",\"NM_001040142.2(SCN2A):c.605C>T (p.Ala202Val)\"],\n",
    "    \"N503K*19\": [\"c.1508dup\", \"NM_001040142.2:c.1508dup NP_001035232.1:p.(Asn503LysfsTer19)\"],\n",
    "    \"L611V*35\": [\"c.1831_1832del\", \"NM_001040142.2:c.1831_1832del  NP_001035232.1:p.(Leu611ValfsTer35)\"],\n",
    "    \"E440R*20\": [\"c.1318_1349del\", \"NM_001040142.2:c.1318_1349del  NP_001035232.1:p.(E440Rfs*20)\"],\n",
    "    \"F1861L*40\": [\"c.5583del\", \"NM_001040142.2:c.5583del  NP_001035232.1:p.(F1861Lfs*40)\"],\n",
    "    \"M1797I*5\": [\"c.5387_5390dup\", \"NM_001040142.2:c.5387_5390dup NP_001035232.1:p.(M1797Ifs*5)\"],\n",
    "    \"Q1803G\": [\"c.5408A>G\", \"NM_001040142.2:c.5408A>G  NP_001035232.1:p.(E1803G)\"],\n",
    "    \"Q169G\": [\"c.506A>G\", \"NM_001040142.2:c.506A>G  NP_001035232.1:p.(E169G)\"],\n",
    "    \"I1537_M1538delinsSI\": [\"c.4610_4614delinsGCATC\", \"NM_001040142.2:c.4610_4614delinsGCATC NP_001035232.1:p.(I1537_M1538delinsSI)\"],\n",
    "     \"T784C*45\" : [\"c.2350_2365delinsTGTACTATCCAACAGATACT\", \"NM_001040142.2:c.2350_2365delinsTGTACTATCCAACAGATACT  NP_001035232.1:p.(T784Cfs*45)\"],\n",
    "    \"I874M*5\": [\"c.2622_2631del\", \"NM_001040142.2(SCN2A):c.2622_2631del NP_001035232.1:p.(Ile874MetfsTer5) NP_001035232.1:p.(I874Mfs*5)\"],\n",
    "    \"K1387S*4\": [\"c.4160_4161del\", \"NM_001040142.2(SCN2A):c.4160_4161del (p.Lys1387fs)\"],\n",
    "    \"I1615R*47\": [\"c.4844_4845del\", \"NM_001040142.2(SCN2A):c.4844_4845del (p.Ile1615fs) \"]\n",
    "\n",
    "}\n",
    "\n",
    "malformed = { \n",
    "    \"G999L\",\n",
    "    \"V1261M\",\n",
    "    \"V1263M\",\n",
    "    \"V1528C*7\",\n",
    "    \"Q1904R*22\",\n",
    "    \"Q430G\",\n",
    "    \"L392L\", # splice but no nt available\n",
    "    \"F1599C*14\", # original article ambiguous\n",
    "    \"N361T*21\", # cannot find\n",
    "    \"T435I*5\", # cannot find\n",
    "    \"C1170V*15\", # cannot find\n",
    "    \"T1711L*8\",     # cannot find\n",
    "    \"K1260E;K1260Q\",\n",
    "    \"[R1882G;G1522A]\",\n",
    "    \"Del(exons_18-27)\",\n",
    "    \"missense\",\n",
    "    \"2q24.3 duplication\"           \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c52ef911",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Got 396 mappings\n"
     ]
    }
   ],
   "source": [
    "mappings = []\n",
    "for famID, var in fam_to_var_d.items():\n",
    "    #print(famID, var)\n",
    "    if var.startswith(\"c.\"):\n",
    "        mane = var\n",
    "    elif var == '4551+1C>A':\n",
    "        mane = \"c.4551+1C>A\"\n",
    "    elif var == \"c.2379+1G>A\":\n",
    "        mane = \"c.2388+1G>A\"  # corrected from PMID:26647175 on basis of chromosomal position\n",
    "    elif var in protein_to_NM_MANE_d:\n",
    "        mane = protein_to_NM_MANE_d.get(var)\n",
    "    elif var in manual_lookups:\n",
    "        mane = manual_lookups.get(var)[0]\n",
    "    elif var in  malformed:\n",
    "        continue\n",
    "    else:\n",
    "        print(f\"Could not find {var}\")\n",
    "        continue\n",
    "    m = {'famID' : famID, 'var': var, 'MANE': mane}\n",
    "    mappings.append(m)\n",
    "print(f\"Got {len(mappings)} mappings\")\n",
    "mapping_df = pd.DataFrame(mappings)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d30a7016",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>famID</th>\n",
       "      <th>var</th>\n",
       "      <th>MANE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>fam1</td>\n",
       "      <td>A202V</td>\n",
       "      <td>c.605C&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>fam10</td>\n",
       "      <td>V423L</td>\n",
       "      <td>c.1267G&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>fam100</td>\n",
       "      <td>L501*</td>\n",
       "      <td>c.1501_1502delinsTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>fam101</td>\n",
       "      <td>N503K*19</td>\n",
       "      <td>c.1508dup</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>fam102</td>\n",
       "      <td>R583*</td>\n",
       "      <td>c.1747C&gt;T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    famID       var                 MANE\n",
       "0    fam1     A202V             c.605C>T\n",
       "1   fam10     V423L            c.1267G>T\n",
       "2  fam100     L501*  c.1501_1502delinsTA\n",
       "3  fam101  N503K*19            c.1508dup\n",
       "4  fam102     R583*            c.1747C>T"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b3850ea2",
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping_df.to_csv(\"variant_mapping.tsv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fc3af27",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
